core/src/consensus.rs

```diff
@@ -651,6 +945,144 @@ impl Tower {
+        fs::create_dir_all(&filename.parent().unwrap())?;
+
+        let file = File::open(&filename)?;
+        let mut stream = BufReader::new(file);
+
+        let saved_tower: SavedTower = bincode::deserialize_from(&mut stream)?;
+        if !saved_tower.verify(node_pubkey) {
+            return Err(TowerError::InvalidSignature);
+        }
+        let mut tower = saved_tower.deserialize()?;
+        tower.path = filename;
+        tower.tmp_path = Self::get_tmp_filename(&tower.path);
+
+        // check that the tower actually belongs to this node
+        if &tower.node_pubkey != node_pubkey {
+            return Err(TowerError::WrongTower(format!(
+                "node_pubkey is {:?} but found tower for {:?}",
+                node_pubkey, tower.node_pubkey
+            )));
+        }
+        Ok(tower)
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum TowerError {
+    #[error("IO Error: {0}")]
+    IOError(#[from] std::io::Error),
+
+    #[error("Serialization Error: {0}")]
+    SerializeError(#[from] bincode::Error),
+
+    #[error("The signature on the saved tower is invalid")]
+    InvalidSignature,
+
+    #[error("The tower does not match this validator: {0}")]
+    WrongTower(String),
+
+    #[error(
+        "The tower is too old: \
+        newest slot in tower ({0}) << oldest slot in available history ({1})"
+    )]
+    TooOldTower(Slot, Slot),
+
+    #[error(
+        "The slot history is too old: \
+        oldest slot in tower ({0}) >> newest slot in available history ({1})"
+    )]
+    TooOldSlotHistory(Slot, Slot),
+
+    #[error("The tower is inconsistent with slot history: {0}")]
+    InconsistentWithSlotHistory(String),
+}
+
+#[frozen_abi(digest = "Gaxfwvx5MArn52mKZQgzHmDCyn5YfCuTHvp5Et3rFfpp")]
+#[derive(Default, Clone, Serialize, Deserialize, Debug, PartialEq, AbiExample)]
+pub struct SavedTower {
+    signature: Signature,
+    data: Vec<u8>,
+}
+
+impl SavedTower {
+    pub fn new<T: Signer>(tower: &Tower, keypair: &Arc<T>) -> Result<Self> {
+        let data = bincode::serialize(tower)?;
+        let signature = keypair.sign_message(&data);
+        Ok(Self { data, signature })
+    }
+
+    pub fn verify(&self, pubkey: &Pubkey) -> bool {
+        self.signature.verify(pubkey.as_ref(), &self.data)
+    }
+
+    pub fn deserialize(&self) -> Result<Tower> {
+        bincode::deserialize(&self.data).map_err(|e| e.into())
+    }
+}
+
+// Given an untimely crash, tower may have roots that are not reflected in blockstore because
+// `ReplayState::handle_votable_bank()` saves tower before setting blockstore roots
+pub fn reconcile_blockstore_roots_with_tower(
+    tower: &Tower,
+    blockstore: &Blockstore,
+) -> blockstore_db::Result<()> {
+    if let Some(tower_root) = tower.root() {
+        let last_blockstore_root = blockstore.last_root();
+        if last_blockstore_root < tower_root {
+            let new_roots: Vec<_> = blockstore
+                .slot_meta_iterator(last_blockstore_root + 1)?
```

Comment:
The problem with `slot_meta_iterator` is that it will iterate through any slots in blockstore in order, even if those slots are not necessarily descended from each other.

Because we only want to set roots for parents of `tower_root`, we should look up the SlotMeta for `tower_root` and traverse its parents until we reach `last_blockstore_root`, and set all those parents as roots.

I think the proper iterator here would be `AncestorIterator`, similar to here: https://github.com/solana-labs/solana/blob/master/core/src/repair_weight.rs#L408

Reply:
Hooray! Thanks for spotting this before we land this.... This bug was there since #9902... 

Done in https://github.com/solana-labs/solana/pull/10718/commits/40c7bb7674ac0461a4d2826d67927b8681fb21eb